File: TDD/TDD Coding Practice Arithmetic/Double Stack Algorithm/Code/Python/oo/parser/expression.py
from doublestackprocessor import DoubleStackProcessor

class Expression(object):
    def __init__(self, expr, context=None):
        if context == None:
            self.ctx = Context(expr, 0)
        else:
            self.ctx = context

        self.dsp = DoubleStackProcessor()
        self.parser = Parser(self.ctx)

    def eval(self):
        while self.parser.expr_not_end():
            item = self.parser.next_item()
            if '(' == item:
                self.dsp.push_operand(Expression(self.ctx.expr, self.ctx).eval())
            elif ')' == item:
                return self.dsp.result()
            else:
                self.dsp.process(item)
        
        return self.dsp.result()


class Parser(object):
    def __init__(self, ctx):
        self.ctx = ctx

    def expr_not_end(self):
        return self.ctx.curidx < len(self.ctx.expr)

    def next_item(self):
        c = self._next_char()
        return c + self._subsequence(c)

    def _next_char(self):
        c = self.ctx.expr[self.ctx.curidx]
        self.ctx.curidx += 1
        return c

    def _subsequence(self, c):
        if c.isdigit() or self._isnegative(c):
            return self._next_operand()

        if self._isopchar(c):
            return self._next_operator()

        return ''

    def _next_operand(self):
        return self._next_item(self._isdigitchar)

    def _next_operator(self):
        return self._next_item(self._isopchar)

    def _next_item(self, pred):
        result = ''
        while pred(self._peek_char()):
            result += self._next_char()
        return result

    def _peek_char(self):
        return self.ctx.expr[self.ctx.curidx] if self.expr_not_end() else ''

    def _prev_char(self):
        return self.ctx.expr[self.ctx.curidx - 2]

    def _isdigitchar(self, c):
        return c.isdigit()

    def _isopchar(self, c):
        return c in '+-*/%'

    def _isnegative(self, c):
        return c == '-' and (self.ctx.curidx == 1 or self._prev_char() == '(')


class Context(object):
    def __init__(self, expr, curidx):
        self.expr = expr
        self.curidx = curidx

File: TDD/TDD Coding Practice Arithmetic/Double Stack Algorithm/Code/Java/src/fayelab/tdd/arithmetic/doublestack/parser/Expression.java
package fayelab.tdd.arithmetic.doublestack.parser;

import java.util.function.Predicate;

public class Expression
{
    private Context ctx;
    private DoubleStackProcessor dsp;
    private Parser parser;

    public Expression(String expr)
    {
        this(new Context(expr, 0));
    }

    Expression(Context ctx)
    {
        this.ctx = ctx;
        this.dsp = new DoubleStackProcessor();
        this.parser = new Parser(ctx);
    }

    public int eval()
    {
        while(parser.exprNotEnd())
        {
            String item = parser.nextItem();

            if("(".equals(item))
            {
                dsp.pushOperand(new Expression(ctx).eval());
            }
            else if(")".equals(item))
            {
                return dsp.result();
            }
            else
            {
                dsp.process(item);
            }
        }

        return dsp.result();
    }
}

class Parser
{
    private Context ctx;

    Parser(Context ctx)
    {
        this.ctx = ctx;
    }

    boolean exprNotEnd()
    {
        return ctx.curIdx < ctx.expr.length();
    }

    String nextItem()
    {
        char c = nextChar();
        return c + subsequence(c);
    }

    private char nextChar()
    {
        char c = ctx.expr.charAt(ctx.curIdx);
        ctx.curIdx += 1;
        return c;
    }

    private String subsequence(char c)
    {
        if(isDigit(c) || isNegative(c))
        {
            return nextOperand();
        }

        if(isOpChar(c))
        {
            return nextOperator();
        }

        return "";
    }

    private String nextOperand()
    {
        return nextItem(this::isDigit);
    }

    private String nextOperator()
    {
        return nextItem(this::isOpChar);
    }

    private String nextItem(Predicate<Character> pred)
    {
        String result = "";
        while(pred.test(peekChar()))
        {
            result += nextChar();
        }
        return result;
    }

    private char peekChar()
    {
        return exprNotEnd() ? ctx.expr.charAt(ctx.curIdx) : Character.MIN_VALUE;
    }

    private char prevChar()
    {
        return ctx.expr.charAt(ctx.curIdx - 2);
    }

    private boolean isDigit(char c)
    {
        return '0' <= c && c <= '9';
    }

    private boolean isOpChar(char c)
    {
        return "+-*/%".indexOf(c) > -1;
    }

    private boolean isNegative(char c)
    {
        return c == '-' && (ctx.curIdx == 1 || prevChar() == '(');
    }
}

class Context
{
    String expr;
    int curIdx;

    Context(String expr, int curIdx)
    {
        this.expr = expr;
        this.curIdx = curIdx;
    }
}
